src/util/pointer_offset_size.cpp

```diff
@@ -676,67 +678,80 @@ exprt build_sizeof_expr(
-    const auto &st=to_struct_type(source_type);
-    const struct_typet::componentst &components=st.components();
-    member_offset_iterator offsets(st, ns);
-    while(offsets->first<components.size() &&
-          offsets->second!=-1 &&
-          offsets->second<=offset)
+    const struct_typet &struct_type = to_struct_type(source_type);
+
+    mp_integer m_offset_bits = 0;
+    for(const auto &component : struct_type.components())
     {
-      auto nextit=offsets;
-      ++nextit;
-      if((offsets->first+1)==components.size() || offset<nextit->second)
+      const auto m_size = pointer_offset_bits(component.type(), ns);
+      if(!m_size.has_value())
+        return nil_exprt();
+
+      if(
+        offset * 8 >= m_offset_bits && m_offset_bits % 8 == 0 &&
+        offset * 8 + *target_size <= m_offset_bits + *m_size)
       {
-        // This field might be, or might contain, the answer.
-        result=
-          member_exprt(
-            result,
-            components[offsets->first].get_name(),
-            components[offsets->first].type());
-        return
-          get_subexpression_at_offset(
-            result, offset-offsets->second, target_type, ns);
+        const member_exprt member(expr, component.get_name(), component.type());
+        return get_subexpression_at_offset(
+          member, offset - m_offset_bits / 8, target_type_raw, ns);
       }
-      ++offsets;
+
+      m_offset_bits += *m_size;
     }
-    return nil_exprt();
   }
   else if(source_type.id()==ID_array)
   {
-    // A cell of the array might be, or contain, the subexpression
-    // we're looking for.
-    const auto &at=to_array_type(source_type);
-
-    auto elem_size = pointer_offset_size(at.subtype(), ns);
-
-    if(!elem_size.has_value())
-      return nil_exprt();
+    const array_typet &array_type = to_array_type(source_type);
 
-    mp_integer cellidx = offset / (*elem_size);
+    const auto elem_size = pointer_offset_bits(array_type.subtype(), ns);
 
-    if(cellidx < 0 || !cellidx.is_long())
+    // no arrays of non-byte, zero-, or unknown-sized objects
+    if(!elem_size.has_value() || *elem_size == 0 || *elem_size % 8 != 0)
       return nil_exprt();
 
-    offset = offset % (*elem_size);
+    if(*target_size <= *elem_size)
+      return get_subexpression_at_offset(
+        index_exprt(
+          expr, from_integer(offset / (*elem_size / 8), index_type())),
+        offset % (*elem_size / 8),
+        target_type_raw,
+        ns);
+  }
 
-    result=index_exprt(result, from_integer(cellidx, unsignedbv_typet(64)));
-    return get_subexpression_at_offset(result, offset, target_type, ns);
+  const auto source_size = pointer_offset_bits(source_type, ns);
+  if(
+    offset == 0 && source_size.has_value() && *source_size == *target_size &&
+    source_type.id() == ID_pointer && target_type_raw.id() == ID_pointer)
+  {
+    return typecast_exprt(expr, target_type_raw);
   }
   else
-    return nil_exprt();
+    return byte_extract_exprt(
```

Comment:
Do we have a better solution than byte-extract for e.g. pointer -> unsigned long? Does `typecast_exprt` do that sort of reinterpret?

Reply:
I've now put a call to the simplifier in here (replacing the entire `if`/`else`) to avoid repeating logic. If pointer <-> unsigned long is deemed ok to do via typecast then we should teach the simplifier about that, I'd say.